packages/utilities/src/lib/mergeDefault.ts

```diff
@@ -1,27 +1,61 @@
 import { deepClone } from './deepClone';
 import { isObject } from './isObject';
-
 import type { DeepRequired } from './utilityTypes';
 
 // eslint-disable-next-line @typescript-eslint/consistent-type-definitions, @typescript-eslint/ban-types
 type NonNullObject = {};
 
 /**
- * Sets default properties on an object that aren't already specified.
- * @since 0.5.0
- * @param def Default properties
- * @param given Object to assign defaults to
+ * Deep merges 2 objects. Properties from the second parameter are applied to the first.
+ * @remark `overwrites` is also mutated!
+ * @remark If the value of a key in `overwrites` is `undefined` then the value of that same key in `base` is used instead!
+ * @remark This is essentially `{ ...base, ...overwrites }` but recursively
+ * @since 1.0.0
+ * @param base Base object
+ * @param overwrites Overwrites to apply
+ * @example
+ * ```ts
+ * const base = { a: 0, b: 1 };
+ * const overwrites = {}; // will be { a: 0, b: 1 } after merge
+ * mergeDefault(base, overwrites) // { a: 0, b: 1 }
+ * ```
+ * @example
+ * ```ts
+ * const base = { a: 0, b: 1 };
+ * const overwrites = { a: 2, i: 3 };
+ * mergeDefault(base, overwrites) // { a: 2, i: 3, b: 1 };
+ * ```
+ * @example
+ * ```ts
+ * const base = { a: 0, b: 1 };
+ * const overwrites = { a: null };
+ * mergeDefault(base, overwrites) // { a: null, b: 1 };
+ * ```
+ * @example
+ * ```ts
+ * const base = { a: 0, b: 1 };
+ * const overwrites = { a: undefined };
+ * mergeDefault(base, overwrites) // { a: 0, b: 1 };
+ * ```
+ * @example
+ * ```ts
+ * const base = { a: null };
+ * const overwrites = { a: { b: 5 } };
+ * mergeDefault(base, overwrites) // { a: { b: 5 } };
+ * ```
  */
-export function mergeDefault<A extends NonNullObject, B extends Partial<A>>(defaults: A, given?: B): DeepRequired<A & B> {
-	if (!given) return deepClone(defaults) as DeepRequired<A & B>;
-	for (const [key, value] of Object.entries(defaults)) {
-		const givenValue = Reflect.get(given, key);
+export function mergeDefault<A extends NonNullObject, B extends Partial<A>>(base: A, overwrites?: B): DeepRequired<A & B> {
+	// If no overwrites are specified then deep clone the base
+	if (!overwrites) return deepClone(base) as DeepRequired<A & B>;
+
+	for (const [key, value] of Object.entries(base)) {
+		const givenValue = Reflect.get(overwrites, key);
 		if (typeof givenValue === 'undefined') {
-			Reflect.set(given, key, deepClone(value));
+			Reflect.set(overwrites, key, deepClone(value));
 		} else if (isObject(givenValue)) {
-			Reflect.set(given, key, mergeDefault(value as NonNullObject, givenValue));
+			Reflect.set(overwrites, key, mergeDefault((value ?? {}) as NonNullObject, givenValue));
```

Comment:
Why are we `?? {}`?

Reply:
![image](https://user-images.githubusercontent.com/4019718/95691513-00646b80-0c20-11eb-9132-d5fe73c4d0df.png)

Reply:
Bug that was carried over from Klasa

Reply:
I don't know, it will still break if I do
```ts
mergeDefault({ a: 1 }, { a: { b: true } })
```
won't it?

Reply:
it won't